From 927ff6cf678a6a382d81f345eb57b19274a57292 Mon Sep 17 00:00:00 2001
From: Patrick Gaskin <patrick@pgaskin.net>
Date: Sun, 1 Oct 2023 23:05:33 -0400
Subject: [PATCH 4/9] Implement key repeat for extension keyboard

Allows keys to be long-pressed (typically 500ms) to start repeating
them at 75ms. Doesn't apply to modifier keys.

The way I did this is somewhat ugly and breaks encapsulation, but
it's the simplest way to do it in a mostly-additive way to keep this
maintainable.

Updating:
- Find the class extending SimpleOnGestureListener (try: ag 'super Landroid/view/GestureDetector\$SimpleOn
GestureListener').
- Test to ensure key repeat works for non-modifier keys and doesn't for modifiers.

diff --git a/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali b/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali
index a431ccf..6bf8883 100644
--- a/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali
+++ b/smali/com/realvnc/viewer/android/ui/KeyboardKey.smali
@@ -4,13 +4,13 @@
 
 
 # instance fields
-.field private a:I
+.field private a:I # key state (up=0, down=1, locked=2)
 
 .field private b:Lx2/t;
 
 .field private c:Lx2/s;
 
-.field private d:I
+.field private d:I # whether key can be locked (no=0, 1=yes)
 
 .field private e:Landroid/view/GestureDetector;
 
@@ -608,6 +608,8 @@
 
     invoke-virtual {v0, v1, v2, v3}, Lx/e;->h(Ljava/lang/Runnable;J)V
 
+    invoke-direct {p0, p1}, Lcom/realvnc/viewer/android/ui/KeyboardKey;->onTouchEventCheckRepeat(Landroid/view/MotionEvent;)V
+
     iget-object v0, p0, Lcom/realvnc/viewer/android/ui/KeyboardKey;->e:Landroid/view/GestureDetector;
 
     invoke-virtual {v0, p1}, Landroid/view/GestureDetector;->onTouchEvent(Landroid/view/MotionEvent;)Z
@@ -616,3 +618,19 @@
 
     return p1
 .end method
+
+.method private onTouchEventCheckRepeat(Landroid/view/MotionEvent;)V
+    .locals 1
+
+    invoke-virtual {p1}, Landroid/view/MotionEvent;->getAction()I
+    move-result p1
+
+    const/4 v0, 0x1 # MotionEvent.ACTION_DOWN (0), MotionEvent.ACTION_UP (1)
+    if-gt p1, v0, :out
+
+    iget-object p0, p0, Lcom/realvnc/viewer/android/ui/KeyboardKey;->f:Lx2/q;
+    invoke-virtual {p0}, Lx2/q;->stopKeyRepeat()V
+
+    :out
+    return-void
+.end method
diff --git a/smali/x2/q$KeyRepeatRunnable.smali b/smali/x2/q$KeyRepeatRunnable.smali
new file mode 100644
index 0000000..22cd03c
--- /dev/null
+++ b/smali/x2/q$KeyRepeatRunnable.smali
@@ -0,0 +1,29 @@
+.class final Lx2/q$KeyRepeatRunnable;
+.super Ljava/lang/Object;
+.implements Ljava/lang/Runnable;
+
+.field private keyRepeatHandler:Landroid/os/Handler;
+.field public key:Lcom/realvnc/viewer/android/ui/KeyboardKey;
+
+.method constructor <init>(Landroid/os/Handler;)V
+    .locals 0
+    invoke-direct {p0}, Ljava/lang/Object;-><init>()V
+    iput-object p1, p0, Lx2/q$KeyRepeatRunnable;->keyRepeatHandler:Landroid/os/Handler;
+    return-void
+.end method
+
+.method public final run()V
+    .locals 3
+
+    iget-object v0, p0, Lx2/q$KeyRepeatRunnable;->key:Lcom/realvnc/viewer/android/ui/KeyboardKey;
+    if-eqz v0, :out
+
+    invoke-virtual {v0}, Lcom/realvnc/viewer/android/ui/KeyboardKey;->n()V
+
+    iget-object v0, p0, Lx2/q$KeyRepeatRunnable;->keyRepeatHandler:Landroid/os/Handler;
+    const-wide/16 v1, 75 # delay
+    invoke-virtual {v0, p0, v1, v2}, Landroid/os/Handler;->postDelayed(Ljava/lang/Runnable;J)Z
+
+    :out
+    return-void
+.end method
diff --git a/smali/x2/q.smali b/smali/x2/q.smali
index cb93291..08354b3 100644
--- a/smali/x2/q.smali
+++ b/smali/x2/q.smali
@@ -5,14 +5,41 @@
 
 # instance fields
 .field private a:Lcom/realvnc/viewer/android/ui/KeyboardKey;
+.field private keyRepeatHandler:Landroid/os/Handler;
+.field private keyRepeatRunnable:Lx2/q$KeyRepeatRunnable;
+
+
+# direct methods
+.method public constructor <init>()V
+    .locals 2
+
+    invoke-direct {p0}, Landroid/view/GestureDetector$SimpleOnGestureListener;-><init>()V
+
+    invoke-static {}, Landroid/os/Looper;->getMainLooper()Landroid/os/Looper;
+    move-result-object v1
+
+    new-instance v0, Landroid/os/Handler;
+    invoke-direct {v0, v1}, Landroid/os/Handler;-><init>(Landroid/os/Looper;)V
+    iput-object v0, p0, Lx2/q;->keyRepeatHandler:Landroid/os/Handler;
+    move-object v1, v0
+
+    new-instance v0, Lx2/q$KeyRepeatRunnable;
+    invoke-direct {v0, v1}, Lx2/q$KeyRepeatRunnable;-><init>(Landroid/os/Handler;)V
+    iput-object v0, p0, keyRepeatRunnable:Lx2/q$KeyRepeatRunnable;
+
+    return-void
+.end method
 
 
 # virtual methods
 .method public final a(Lcom/realvnc/viewer/android/ui/KeyboardKey;)V
-    .locals 0
+    .locals 1
 
     iput-object p1, p0, Lx2/q;->a:Lcom/realvnc/viewer/android/ui/KeyboardKey;
 
+    iget-object v1, p0, Lx2/q;->keyRepeatRunnable:Lx2/q$KeyRepeatRunnable;
+    iput-object p1, v1, Lx2/q$KeyRepeatRunnable;->key:Lcom/realvnc/viewer/android/ui/KeyboardKey;
+
     return-void
 .end method
 
@@ -47,3 +74,30 @@
 
     return p1
 .end method
+
+.method public final onLongPress(Landroid/view/MotionEvent;)V
+    .locals 2
+
+    iget-object v0, p0, Lx2/q;->a:Lcom/realvnc/viewer/android/ui/KeyboardKey;
+    invoke-virtual {v0}, Lcom/realvnc/viewer/android/ui/KeyboardKey;->f()I # getter for KeyboardKey.d (which is 1 if key can be locked down)
+    move-result v0
+    if-eqz v0, :repeat # if key can't be locked (i.e., is not a modifier), start repeating
+    return-void
+
+    :repeat
+    iget-object v0, p0, Lx2/q;->keyRepeatHandler:Landroid/os/Handler;
+    iget-object v1, p0, Lx2/q;->keyRepeatRunnable:Lx2/q$KeyRepeatRunnable;
+    invoke-virtual {v0, v1}, Landroid/os/Handler;->post(Ljava/lang/Runnable;)Z
+
+    return-void
+.end method
+
+.method public final stopKeyRepeat()V
+    .locals 2
+
+    iget-object v0, p0, Lx2/q;->keyRepeatHandler:Landroid/os/Handler;
+    iget-object v1, p0, Lx2/q;->keyRepeatRunnable:Lx2/q$KeyRepeatRunnable;
+    invoke-virtual {v0, v1}, Landroid/os/Handler;->removeCallbacks(Ljava/lang/Runnable;)V
+
+    return-void
+.end method
